File: heroscapebuilder.client/src/models/battlegroup.ts
import { PointSystem } from './point-system';
import { Unit } from './unit';

export interface BattlegroupUnit {
    unit: Unit;
    quantity: number;
    /** Copies the owner has in My Army. Only meaningful for the owner. */
    ownedQuantity: number;
    /** The battlegroup uses more copies than My Army currently holds. Only set for the owner. */
    overAllocated: boolean;
}

export interface Battlegroup {
    id: number;
    name: string;
    pointLimit: number;
    /** The point values this Battlegroup is built with; totalPoints is measured in it. */
    pointSystem: PointSystem;
    /** Null/undefined when units from any creator are allowed. */
    creator?: string | null;
    /** The owner's plain-text notes on how to play this Battlegroup. */
    notes?: string | null;
    isShared: boolean;
    /** Only present for the owner. */
    shareId?: string | null;
    totalPoints: number;
    isOwner: boolean;
    needsReview: boolean;
    reviewReasons: string[];
    updatedAt: string;
    units: BattlegroupUnit[];
}

export interface BattlegroupSaveRequest {
    name: string;
    pointLimit: number;
    pointSystem: PointSystem;
    creator?: string | null;
    notes?: string | null;
    units: Array<{ unitId: number; quantity: number }>;
}

File: HeroscapeBuilder.Server/Data/Entities/Battlegroup.cs
namespace HeroscapeBuilder.Server.Data.Entities;

public partial class Battlegroup
{
    public int Id { get; set; }

    public string UserId { get; set; } = null!;

    public string Name { get; set; } = null!;

    public int PointLimit { get; set; }

    /// <summary>
    /// When set, every unit in the battlegroup must come from this creator. Null means any creator.
    /// </summary>
    public string? Creator { get; set; }

    /// <summary>
    /// Free-form, plain-text notes from the owner on how to play the battlegroup.
    /// </summary>
    public string? Notes { get; set; }

    public bool IsShared { get; set; }

    public Guid ShareId { get; set; }

    public DateTime CreatedAt { get; set; }

    public DateTime UpdatedAt { get; set; }

    public virtual ApplicationUser User { get; set; } = null!;

    public virtual ICollection<BattlegroupUnit> BattlegroupUnits { get; set; } = new List<BattlegroupUnit>();
}
